ethereum/core/src/main/java/org/hyperledger/besu/ethereum/transaction/BlockSimulator.java

```diff
@@ -114,151 +125,172 @@ public List<BlockSimulationResult> process(
   /**
    * Processes a list of BlockStateCalls sequentially, collecting the results.
    *
-   * @param header The block header for all simulations.
-   * @param blockSimulationParameter The blockSimulationParameter containing the block state calls.
+   * @param blockHeader The block header for all simulations.
+   * @param simulationParameter The BlockSimulationParameter containing the block state calls.
    * @param worldState The initial MutableWorldState to start with.
    * @return A list of BlockSimulationResult objects from processing each BlockStateCall.
    */
   public List<BlockSimulationResult> process(
-      final BlockHeader header,
-      final BlockSimulationParameter blockSimulationParameter,
+      final BlockHeader blockHeader,
+      final BlockSimulationParameter simulationParameter,
       final MutableWorldState worldState) {
-    List<BlockSimulationResult> simulationResults = new ArrayList<>();
+    List<BlockSimulationResult> results = new ArrayList<>();
 
-    // Fill gaps between blocks
+    // Fill gaps between blocks and set the correct block number and timestamp
     List<BlockStateCall> blockStateCalls =
-        fillBlockStateCalls(blockSimulationParameter.getBlockStateCalls(), header);
+        fillBlockStateCalls(simulationParameter.getBlockStateCalls(), blockHeader);
 
-    for (BlockStateCall blockStateCall : blockStateCalls) {
-      BlockSimulationResult simulationResult =
+    BlockHeader currentBlockHeader = blockHeader;
+    for (BlockStateCall stateCall : blockStateCalls) {
+      BlockSimulationResult result =
           processBlockStateCall(
-              header, blockStateCall, worldState, blockSimulationParameter.isValidation());
-      simulationResults.add(simulationResult);
+              currentBlockHeader, stateCall, worldState, simulationParameter.isValidation());
+      results.add(result);
+      currentBlockHeader = result.getBlock().getHeader();
```

Comment:
is it correct to say that previously the block header was always the same for each call to `processBlockStateCall` while now it is update on every loop?

Reply:
Yes, it is correct. The next block simulation needs to be based on the outcome of the previous one. I will update the PR description.